rpc/http.go

```diff
@@ -313,3 +97,99 @@ func (client *httpClient) Close() {
 func (client *httpClient) SupportedModules() (map[string]string, error) {
 	return SupportedModules(client)
 }
+
+// httpStream reads and writes HTTP requests and responses and can be used
+// as the ReadWriteCloser for the ServerCodec.
+type httpStream struct {
+	mu              sync.Mutex
+	request         *http.Request
+	firstRead     bool
+	response        http.ResponseWriter
+	responseWritten chan struct{}
+}
+
+// Read the body of the next HTTP request.
+func (rw *httpStream) Read(p []byte) (int, error) {
+	n, err := rw.request.Body.Read(p)
+	if n == 0 && err == io.EOF && !rw.firstRead { // request already read
+		<-rw.responseWritten // wait for response to be written (http is sync)
+		return 0, io.EOF     // indicate that the request has been read
+	}
+	rw.firstRead = false
+	return n, err
+}
+
+// Write the given buffer as a HTTP response.
+// Note: this will write the headers which won't allow the response to be
+// written in multiple chunks.
+func (rw *httpStream) Write(p []byte) (int, error) {
+	rw.mu.Lock()
+	defer rw.mu.Unlock()
+	if rw.responseWritten != nil {
+		n, err := rw.response.Write(p)
+		close(rw.responseWritten)
+		rw.responseWritten = nil
+		return n, err
+	}
+	return 0, errors.New("response already written")
+}
+
+// Close indicates that the HTTP response has been sent.
+func (rw *httpStream) Close() error {
+	rw.mu.Lock()
+	defer rw.mu.Unlock()
+	if rw.responseWritten != nil {
+		close(rw.responseWritten)
+		rw.responseWritten = nil
+	}
+	return nil
+}
+
+// newJSONHTTPHandler creates a HTTP handler that will parse incoming JSON requests,
+// send the request to the given API provider and sends the response back to the caller.
+func newJSONHTTPHandler(srv *Server) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.ContentLength > maxHTTPRequestContentLength {
+			http.Error(w,
+				fmt.Sprintf("content length too large (%d>%d)", r.ContentLength, maxHTTPRequestContentLength),
+				http.StatusRequestEntityTooLarge)
+			return
+		}
+
+		// turn message based HTTP into a stream that can be used by the ServerCodec
+		rw := &httpStream{
+			request:         r,
+			response:        w,
+			firstRead:       true,
+			responseWritten: make(chan struct{}),
+		}
+
+		codec := NewJSONCodec(rw)
+		w.Header().Set("content-type", "application/json")
+
+		go srv.ServeCodec(codec)
+
+		<-rw.responseWritten
```

Comment:
I'm not entirely sure I understand this construct here. If I get it correctly, the reason you have this channel is to ensure `ServeCodec` finishes before returning. But wouldn't it make more sense to simply not start the serve in a new goroutine but rather inside this handler. That way you wouldn't need the blocking channel at all.

The reason I'm a bit afraid of this construct is because quite a few things can go wrong:

 * The serving code never signals back the handler, leaking out goroutines
 * In both `Write` and `Close` you close the channel and immediately afterwards set it to nil. This can cause deadlocks if this handler here doesn't reach the channel read fast enough, since by the time it reaches it the channel will be nil and read will block. Similarly there is a race condition accessing this channel, if other code sets it to nil.

Is there really a valid need for this channel?